camus-kafka-coders/src/main/java/io/confluent/camus/etl/kafka/coders/AvroMessageDecoder.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.util.Utf8;
+import org.apache.hadoop.io.Text;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Properties;
+
+import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+import io.confluent.kafka.schemaregistry.client.SchemaMetadata;
+
+public class AvroMessageDecoder extends MessageDecoder<byte[], Record> {
+  private static final byte MAGIC_BYTE = 0x0;
+  private static final int idSize = 4;
+  private static final String SCHEMA_REGISTRY_URL = "schema.registry.url";
+  private static final String MAX_SCHEMAS_PER_SUBJECT = "max.schemas.per.subject";
+  private static final String DEFAULT_MAX_SCHEMAS_PER_SUBJECT = "1000";
+  private static final String IS_NEW_PRODUCER = "is.new.producer";
+  private static final Logger logger = Logger.getLogger(AvroMessageDecoder.class);
+  protected DecoderFactory decoderFactory;
+  private SchemaRegistryClient schemaRegistry;
+  private final Schema.Parser parser = new Schema.Parser();
+  private Schema latestSchema;
+  private int latestVersion;
+  private String topic;
+  private boolean isNew;
+
+  public AvroMessageDecoder(SchemaRegistryClient schemaRegistry) {
+    this.schemaRegistry = schemaRegistry;
+  }
+
+  public AvroMessageDecoder() {
+
+  }
+
+  @Override
+  public void init(Properties props, String topicName) {
+    super.init(props, topicName);
+
+    decoderFactory = DecoderFactory.get();
+    if (props == null) {
+      throw new IllegalArgumentException("Missing schema registry url!");
+    }
+    String baseUrl = props.getProperty(SCHEMA_REGISTRY_URL);
+    if (baseUrl == null) {
+      throw new IllegalArgumentException("Missing schema registry url!");
+    }
+    String maxSchemaObject = props.getProperty(
+        MAX_SCHEMAS_PER_SUBJECT, DEFAULT_MAX_SCHEMAS_PER_SUBJECT);
+    if (schemaRegistry == null) {
+      schemaRegistry = new CachedSchemaRegistryClient(baseUrl, Integer.parseInt(maxSchemaObject));
+    }
+    this.isNew = Boolean.parseBoolean(props.getProperty(IS_NEW_PRODUCER, "true"));
+    this.topic = topicName;
+  }
+
+  private ByteBuffer getByteBuffer(byte[] payload) {
+    ByteBuffer buffer = ByteBuffer.wrap(payload);
+    byte magic = buffer.get();
+    logger.debug("MAGIC BYTE" + magic);
+    if (magic != MAGIC_BYTE) {
+      throw new MessageDecoderException("Unknown magic byte!");
+    }
+    return buffer;
+  }
+
+  private String constructSubject(String topic, Schema schema, boolean isNewProducer) {
+    if (isNewProducer) {
+      return topic + "-value";
+    } else {
+      return schema.getName() + "-value";
+    }
+  }
+
+  private Object deserialize(byte[] payload) throws MessageDecoderException {
+    try {
+      ByteBuffer buffer = getByteBuffer(payload);
```

Comment:
Payload can not be null if the producer is using our KafkaAvroSerializer/KafkaAvroEncoder as the message needs to contain a MAGIC_BYTE and an id for the schema.  Also, ByteBuffter.wrap will throw a null pointer exception on null values.